### PR TITLE
Voxel Grid Implementation

### LINKED ISSUE #185: Implement Voxel-grid
https://pointclouds.org/documentation/classpcl_1_1_voxel_grid.html

### PR DESCRIPTION
**Pull request by abdulrahman-1212**
closes #185  
This PR introduces a VoxelGrid implementation for downsampling 3D point clouds in the kornia-3d crate, inspired by the Point Cloud Library (PCL) VoxelGrid filter. The implementation supports efficient point cloud downsampling by grouping points into a 3D voxel grid and computing centroids, with optional downsampling of colors and normals controlled by the downsample_all_data field. 

##Key Features
1. Voxel Grid Implementation: Downsamples points by computing voxel centroids using configurable leaf_size (DVec3). Optionally downsamples colors (averaged, normalized to [0, 255]) and normals (averaged, normalized to unit length) when downsample_all_data is true. Uses HashMap with VoxelData type alias for clarity and efficiency. Includes utility methods: get_min_bound, get_max_bound, get_voxel_index, set_leaf_size, etc.
 
2. tests Covers initialization, downsampling (points-only and with colors/normals), bounds, and edge cases. Verified with "cargo test --package kornia-3d". 
3. Applied glam vectorization
4. Created voxel_grid example 

Please review and provide feedback!

### CODE REVIEW

File: crates/kornia-3d/src/lib.rs [hunk 1]
```diff
@@ -21,3 +21,6 @@ pub mod transforms;
 
 /// 3D vector traits.
 pub mod vector;
+
+/// Voxel Grid
```
Comment (edgarriba) [inline crates/kornia-3d/src/lib.rs]: make this comment more meaningful

File: crates/kornia-3d/src/vector.rs [hunk 2]
```diff
@@ -40,6 +58,24 @@ impl DVec3 {
             z: array[2],
         }
     }
+
+    /// Add two DVec3 vectors.
+    pub fn add(&self, other: &Self) -> Self {
+        Self {
```
Comment (edgarriba) [inline crates/kornia-3d/src/vector.rs]: probably this can be implemented using glam vectorized semantics

File: crates/kornia-3d/src/vector.rs [hunk 3]
```diff
@@ -40,6 +58,24 @@ impl DVec3 {
...
+            x: self.x + other.x,
+            y: self.y + other.y,
+            z: self.z + other.z,
+        }
+    }
+
+    /// Scale a DVec3 vector by a scalar.
+    pub fn scale(&self, factor: f64) -> Self {
+        Self {
```
Comment (edgarriba) [inline crates/kornia-3d/src/vector.rs]: same here, check if that can be vectorized

File: crates/kornia-3d/src/voxel_grid.rs [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    /// Creates a new `VoxelGrid` with the specified leaf size.
+    ///
+    /// # Arguments
+    /// * `leaf_size` - The size of the voxel in x, y, z dimensions.
+    ///
+    /// # Panics
+    /// Panics if any dimension of `leaf_size` is non-positive.
+    pub fn new(leaf_size: DVec3) -> Self {
+        if leaf_size.x <= 0.0 || leaf_size.y <= 0.0 || leaf_size.z <= 0.0 {
+            panic!("Leaf size must be positive in all dimensions");
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs]: use proper error handling

File: crates/kornia-3d/src/voxel_grid.rs [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+            let iy = (point_vec.y / self.leaf_size.y).floor() as i32;
+            let iz = (point_vec.z / self.leaf_size.z).floor() as i32;
+            let key = (ix, iy, iz);
+            let entry = grid.entry(key).or_insert((
+                DVec3::from_array(&[0.0, 0.0, 0.0]),
+                Vec::new(),
+                Vec::new(),
+                0,
+            ));
+            entry.0.x += point_vec.x;
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs]: for the VoxelData use a proper struct so that there are semantics and this code is more readable/maintainable.

also probably use glam vectors to leverage vectorisation in the addition

File: crates/kornia-3d/src/voxel_grid.rs [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+            entry.0.y += point_vec.y;
+            entry.0.z += point_vec.z;
+            entry.3 += 1;
+
+            if self.downsample_all_data {
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs]: something like this would be more compact and rust idiomatic

```rust
if !self.downsample_all_data {
    return;
}
if let Some(&color)  = point_cloud.colors().and_then(|c| c.get(i)) {
    entry.1.push(color);
}
if let Some(&normal) = point_cloud.normals().and_then(|n| n.get(i)) {
    entry.2.push(normal);
}
```

File: crates/kornia-3d/src/voxel_grid.rs [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+                if let Some(normals) = point_cloud.normals() {
+                    if let Some(normal) = normals.get(i) {
+                        entry.2.push(*normal);
+                    }
+                }
+            }
+        }
+
+        // Compute centroids for each voxel
+        let mut points = Vec::new();
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs]: more compact

```rust
let mut points = Vec::new();

let mut colors = point_cloud
    .colors()
    .filter(|_| self.downsample_all_data)
    .map(|_| Vec::new());

let mut normals = point_cloud
    .normals()
    .filter(|_| self.downsample_all_data)
    .map(|_| Vec::new());
```

File: crates/kornia-3d/src/voxel_grid.rs [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+            Some(Vec::new())
+        } else {
+            None
+        };
+
+        for (_key, (sum, color_vec, normal_vec, count)) in grid {
+            if count >= self.min_points_per_voxel {
+                let inv_count = 1.0 / count as f64;
+                let centroid =
+                    DVec3::from_array(&[sum.x * inv_count, sum.y * inv_count, sum.z * inv_count]);
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs]:
```suggestion
                    sum * inv_count;
```
that should work i think

File: crates/kornia-3d/src/voxel_grid.rs [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+                points.push([centroid.x, centroid.y, centroid.z]);
+
+                if self.downsample_all_data {
+                    if let Some(ref mut colors_vec) = colors {
+                        if !color_vec.is_empty() {
+                            let color_sum: [u64; 3] = color_vec.iter().fold([0, 0, 0], |acc, c| {
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs]: leverage DVec

File: crates/kornia-3d/src/voxel_grid.rs [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+    }
+
+    /// Gets the minimum bound of the point cloud.
+    ///
+    /// # Arguments
+    /// * `point_cloud` - The input point cloud.
+    ///
+    /// # Returns
+    /// The minimum x, y, z coordinates as a `DVec3`. Returns `[0.0, 0.0, 0.0]` if empty.
+    pub fn get_min_bound(&self, point_cloud: &PointCloud) -> DVec3 {
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs:187]: more idiomatic
```rust
if point_cloud.points().is_empty() {
    DVec3::from_array(&[0.0; 3])
} else {
    point_cloud
        .points()
        .iter()
        .map(DVec3::from_array)
        .fold(DVec3::from_array(&[f64::INFINITY; 3]), |a, b| a.min(b))
}
```

File: crates/kornia-3d/src/voxel_grid.rs [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+
+    /// Gets the maximum bound of the point cloud.
+    ///
+    /// # Arguments
+    /// * `point_cloud` - The input point cloud.
+    ///
+    /// # Returns
+    /// The maximum x, y, z coordinates as a `DVec3`. Returns `[0.0, 0.0, 0.0]` if empty.
+    pub fn get_max_bound(&self, point_cloud: &PointCloud) -> DVec3 {
+        if point_cloud.points().is_empty() {
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs:208]: check above if that can be adapted

File: crates/kornia-3d/src/voxel_grid.rs [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+    ///
+    /// # Arguments
+    /// * `point` - The input point as `[x, y, z]`.
+    ///
+    /// # Returns
+    /// The voxel index as `(i32, i32, i32)`.
+    pub fn get_voxel_index(&self, point: &[f64; 3]) -> (i32, i32, i32) {
+        let point_vec = DVec3::from_array(point);
+        (
+            (point_vec.x / self.leaf_size.x).floor() as i32,
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs]: vectorize the division at least

File: crates/kornia-3d/src/voxel_grid.rs [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+    /// Sets the voxel grid leaf size.
+    ///
+    /// # Arguments
+    /// * `leaf_size` - The new leaf size in x, y, z dimensions.
+    ///
+    /// # Panics
+    /// Panics if any dimension of `leaf_size` is non-positive.
+    pub fn set_leaf_size(&mut self, leaf_size: DVec3) {
+        if leaf_size.x <= 0.0 || leaf_size.y <= 0.0 || leaf_size.z <= 0.0 {
+            panic!("Leaf size must be positive in all dimensions");
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs]: return proper error

File: crates/kornia-3d/src/voxel_grid.rs [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+        }
+        self.leaf_size = leaf_size;
+    }
+
+    /// Gets the voxel grid leaf size.
+    ///
+    /// # Returns
+    /// The current leaf size as a `DVec3`.
+    pub fn get_leaf_size(&self) -> DVec3 {
+        self.leaf_size.clone()
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs]: why clone

File: crates/kornia-3d/src/voxel_grid.rs [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+    /// * `min_points` - The minimum number of points.
+    pub fn set_min_points_per_voxel(&mut self, min_points: usize) {
+        self.min_points_per_voxel = min_points;
+    }
+
+    /// Gets the minimum number of points required per voxel.
+    ///
+    /// # Returns
+    /// The current minimum number of points.
+    pub fn get_min_points_per_voxel(&self) -> usize {
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs:271]:
```suggestion
    #[inline]
    pub fn get_min_points_per_voxel(&self) -> usize {
```

Comment (edgarriba) [reply crates/kornia-3d/src/voxel_grid.rs:271]: same below for simple getters

File: crates/kornia-3d/src/voxel_grid.rs [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::vector::DVec3;
+
+    #[test]
+    fn test_new_voxel_grid() {
+        let leaf_size = DVec3::from_array(&[1.0, 1.0, 1.0]);
+        let voxel_grid = VoxelGrid::new(leaf_size.clone());
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs]: no need of clone

**Review by edgarriba — CHANGES_REQUESTED** [review]
including a very simple example showing how to load a pointcloud (check other examples doing this) and downsampling with viz in rerun

File: crates/kornia-3d/src/vector.rs [hunk 17]
```diff
@@ -18,10 +19,11 @@ impl Vec3 {
             z: array[2],
         }
     }
+
```
Comment (edgarriba) [inline crates/kornia-3d/src/vector.rs:22]: are applying formatting ? vscode and other should be able to use autoformat

File: crates/kornia-3d/src/voxel_grid.rs [hunk 18]
```diff
@@ -0,0 +1,415 @@
+use std::collections::HashMap;
+use glam::DVec3;
+use crate::pointcloud::PointCloud;
+
+/// Error returned when the leaf size is non-positive in any dimension.
+#[derive(Debug, Clone, Copy)]
+pub struct InvalidLeafSizeError;
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs:7]: use `thiserror` for handle properly the errors

File: crates/kornia-3d/src/voxel_grid.rs [hunk 19]
```diff
@@ -0,0 +1,415 @@
...
+
+impl std::fmt::Display for InvalidLeafSizeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Leaf size must be positive in all dimensions")
+    }
+}
+
+impl std::error::Error for InvalidLeafSizeError {}
+
+/// Data structure for voxel data when downsampling all data.
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs:17]:
```suggestion
/// Data structure to represent a voxel in a 3d voxel grid.
```

File: crates/kornia-3d/src/voxel_grid.rs [hunk 20]
```diff
@@ -0,0 +1,415 @@
...
+    point_sum: DVec3,
+    /// List of colors in the voxel.
+    colors: Vec<[u8; 3]>,
+    /// List of normals in the voxel.
+    normals: Vec<[f64; 3]>,
+    /// Number of points in the voxel.
+    count: usize,
+}
+
+/// A 3D voxel grid for downsampling point clouds.
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs:30]:
```suggestion
/// Data structure to represent a discrete and uniform 3d structure.
```

File: crates/kornia-3d/src/voxel_grid.rs (hunk 19 above)
Comment (edgarriba) [reply crates/kornia-3d/src/voxel_grid.rs:17]: this is not only for downsampling i guess

File: crates/kornia-3d/src/voxel_grid.rs [hunk 21]
```diff
@@ -0,0 +1,415 @@
...
+                let entry = grid.entry(key).or_insert(VoxelData {
+                    point_sum: DVec3::ZERO,
+                    colors: Vec::new(),
+                    normals: Vec::new(),
+                    count: 0,
+                });
+                entry.point_sum += point_vec;
+                entry.count += 1;
+
+                if let Some(colors) = point_cloud.colors() {
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs:85]: more rust idiomatic `let mut colors = point_cloud.colors().map(|_| Vec::new());`
same below

File: crates/kornia-3d/src/voxel_grid.rs [hunk 22]
```diff
@@ -0,0 +1,415 @@
...
+                    points.push([centroid.x, centroid.y, centroid.z]);
+
+                    if let Some(ref mut colors_vec) = colors {
+                        if !voxel_data.colors.is_empty() {
+                            let color_sum: [u64; 3] = voxel_data.colors.iter().fold(
+                                [0, 0, 0],
+                                |acc, c| [acc[0] + c[0] as u64, acc[1] + c[1] as u64, acc[2] + c[2] as u64],
+                            );
+                            colors_vec.push([
+                                (color_sum[0] as f64 * inv_count).round().clamp(0.0, 255.0) as u8,
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs:123]: do the ops in glam them `to_array` to append to the vec

File: crates/kornia-3d/src/voxel_grid.rs [hunk 23]
```diff
@@ -0,0 +1,415 @@
...
+                                (color_sum[1] as f64 * inv_count).round().clamp(0.0, 255.0) as u8,
+                                (color_sum[2] as f64 * inv_count).round().clamp(0.0, 255.0) as u8,
+                            ]);
+                        }
+                    }
+                    if let Some(ref mut normals_vec) = normals {
+                        if !voxel_data.normals.is_empty() {
+                            let normal_sum: [f64; 3] = voxel_data.normals.iter().fold(
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs:131]: same a above

File: crates/kornia-3d/src/voxel_grid.rs [hunk 24]
```diff
@@ -0,0 +1,415 @@
...
+
+    /// Downsamples the input point cloud by grouping points into voxels and computing centroids.
+    ///
+    /// # Arguments
+    /// * `point_cloud` - The input point cloud to downsample.
+    ///
+    /// # Returns
+    /// A new `PointCloud` containing the downsampled points, and optionally colors and normals.
+    pub fn downsample(&mut self, point_cloud: &PointCloud) -> PointCloud {
+        if self.downsample_all_data {
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs:68]: can you split the functions below in: `create_grid_all_data(...)` and `create_grid` ?

Comment (edgarriba) [reply crates/kornia-3d/src/voxel_grid.rs:68]: in that way you could also unit test each

File: crates/kornia-3d/src/voxel_grid.rs [hunk 25]
```diff
@@ -0,0 +1,415 @@
...
+
+    /// Computes the voxel index for a given point.
+    ///
+    /// # Arguments
+    /// * `point` - The input point as `[x, y, z]`.
+    ///
+    /// # Returns
+    /// The voxel index as `(i32, i32, i32)`.
+    #[inline]
+    pub fn get_voxel_index(&self, point: &[f64; 3]) -> (i32, i32, i32) {
```
Comment (edgarriba) [inline crates/kornia-3d/src/voxel_grid.rs:227]: maybe to we could return a fixed array [i32; 3] to easy converstion ?

File: examples/voxel_grid/Cargo.toml
Comment (edgarriba) [inline examples/voxel_grid/Cargo.toml:1]: can you add a README.md similar to the other examples with a screenshot of a working example ?

**Comment by edgarriba** [comment]
Closing this, if there’s still interest, it needs rebase and updates. Thanks !
